File: models/post.go
package models

import "time"

// Post represents a single post within the application.
type Post struct {
	ID        string    `json:"id"`
	UserID    string    `json:"user_id"`
	User      *User     `json:"user" sql:"-"`
	FileID    string    `json:"file_id,omitempty"`
	File      *File     `json:"file" sql:"-"`
	Title     string    `json:"title"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`

	Comments []*Comment `json:"comments" sql:"-"`
}

// Posts is a slice of a post.
type Posts []*Post

func (ps Posts) IDs() []string {
	ids := make([]string, len(ps))
	for i, p := range ps {
		ids[i] = p.ID
	}
	return ids
}

func (ps Posts) Map() map[string]*Post {
	m := make(map[string]*Post, len(ps))
	for i, p := range ps {
		m[p.ID] = ps[i]
	}
	return m
}

File: client/src/components/Post.js
// External Imports
import React, { useState } from 'react';
import axios from 'axios';

// Internal Imports
import useUserState from '../hooks/useUserState';
import Comment from './Comment';

const OPTIONS = { month: "long", day: "numeric", year: 'numeric'};

const Post = (props) => {
    const {user} = useUserState();

    const [error, setError] = useState('');

    const setDate = (date) => {
        return new Date(date).toLocaleString("en", OPTIONS);
    };

    const removePost = (id) => {
        axios.delete(`http://localhost:8080/api/posts/${id}`, { withCredentials: true })
        .then(() => {
            props.submit(true);
        })
        .catch((err) => setError(err));
        props.submit(false);
    };

    return (
        <div className="card card-container">
            {error && (
                <p>An error occurred</p>
            )}
            <div className="card-body">
                <div className="header-container">
                    <div className="fix">
                        <img alt="Profile" className="img-user" src="https://capenetworks.com/static/images/testimonials/user-icon.svg" />
                        <div className="header">
                            <div className="name">{props.post.user.first_name + " " + props.post.user.last_name}</div>
                            <p>{setDate(props.post.created_at)}</p>
                        </div>
                    </div>
                    {user.id === props.post.user_id && (
                        <button className="no-style-btn" onClick={() => removePost(props.post.id)}><i className="fa fa-trash fa-lg" aria-hidden="true"></i></button>
                    )}
                </div>
                <div className="content-container">
                    <h2 className="card-title">{props.post.title}</h2>
                    <p className="card-text">{props.post.content}</p>
                    {props.post.file_id && props.post.file.file_type === "image" && (
                        <img src={props.post.file.s3_url} alt="post" />
                    )}
                    {props.post.file_id && props.post.file.file_type === "video" && (
                        <video className="video" controls>
                            <source src={props.post.file.s3_url} type="video/mp4"></source>
                        </video>
                    )}
                </div>
            </div>
            <Comment post={props.post} comments={props.post.comments} submit={props.submit} setError={setError} />
        </div>
    );
};

export default Post;